File: apps/android/feature/ai/data/src/main/java/com/scribblefit/feature/ai/data/repository/SyncRepositoryImpl.kt
package com.scribblefit.feature.ai.data.repository

import android.content.Context
import androidx.work.BackoffPolicy
import androidx.work.Constraints
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import com.scribblefit.core.database.dao.ExerciseDictionaryDao
import com.scribblefit.core.database.dao.SetDao
import com.scribblefit.core.database.dao.SyncQueueDao
import com.scribblefit.core.database.dao.WorkoutLogDao
import com.scribblefit.core.database.model.SyncQueueEntity
import com.scribblefit.feature.ai.data.worker.SyncWorker
import com.scribblefit.feature.ai.domain.engine.SyncRepository
import com.scribblefit.feature.ai.domain.model.ParsedWorkout
import com.scribblefit.feature.ai.domain.model.SyncItem
import com.scribblefit.feature.ai.domain.model.SyncStatus
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.Json
import java.time.Duration
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import com.scribblefit.core.database.model.SyncStatus as EntitySyncStatus

@Singleton
class SyncRepositoryImpl @Inject constructor(
    @ApplicationContext private val context: Context,
    private val syncQueueDao: SyncQueueDao,
    private val json: Json,
) : SyncRepository {

    internal var workManagerProvider: () -> WorkManager = { WorkManager.getInstance(context) }

    override fun getPendingSyncItems(): Flow<List<SyncItem>> {
        return syncQueueDao.getSyncItemsByStatus(EntitySyncStatus.PENDING).map { entities ->
            entities.map { it.toDomain(json) }
        }
    }

    override fun getAllSyncItems(): Flow<List<SyncItem>> {
        return syncQueueDao.getAllSyncItems().map { entities ->
            entities.map { it.toDomain(json) }
        }
    }

    override suspend fun updateSyncStatus(id: String, status: SyncStatus) {
        syncQueueDao.updateStatus(id, status.toEntity())
    }

    override suspend fun saveParsedWorkout(syncItemId: String, workout: ParsedWorkout) {
        val jsonString = json.encodeToString(workout)
        syncQueueDao.updateParsedResult(syncItemId, EntitySyncStatus.COMPLETED, jsonString)
    }

    override suspend fun enqueueScribble(id: String, rawText: String) {
        val syncItem = SyncQueueEntity(
            id = id,
            type = "SCRIBBLE",
            rawText = rawText,
            status = EntitySyncStatus.PENDING,
            createdAt = System.currentTimeMillis()
        )
        syncQueueDao.upsertSyncItem(syncItem)
        triggerImmediateSync()
    }

    override suspend fun saveFeedItem(id: String, type: String, jsonData: String, status: SyncStatus) {
        val entity = SyncQueueEntity(
            id = id,
            type = type,
            parsedJson = jsonData,
            status = status.toEntity(),
            createdAt = System.currentTimeMillis(),
            rawText = ""
        )
        syncQueueDao.upsertSyncItem(entity)
    }

    override suspend fun deleteSyncItem(id: String) {
        syncQueueDao.deleteById(id)
    }

    private fun triggerImmediateSync() {
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        val syncRequest = OneTimeWorkRequestBuilder<SyncWorker>()
            .setConstraints(constraints)
            .setBackoffCriteria(
                BackoffPolicy.EXPONENTIAL,
                Duration.ofMinutes(1)
            )
            .build()

        workManagerProvider().enqueue(syncRequest)
    }
}

private fun SyncQueueEntity.toDomain(json: Json): SyncItem {
    val parsedWorkout = parsedJson?.let {
        runCatching { json.decodeFromString<ParsedWorkout>(it) }.getOrNull()
    }
    return SyncItem(
        id = id,
        type = type,
        rawText = rawText,
        status = status.toDomain(),
        createdAt = createdAt,
        jsonData = parsedJson,
        parsedResult = parsedWorkout
    )
}

private fun EntitySyncStatus.toDomain(): SyncStatus = when (this) {
    EntitySyncStatus.PENDING -> SyncStatus.PENDING
    EntitySyncStatus.PROCESSING -> SyncStatus.PROCESSING
    EntitySyncStatus.COMPLETED -> SyncStatus.COMPLETED
    EntitySyncStatus.FAILED -> SyncStatus.FAILED
}

private fun SyncStatus.toEntity(): EntitySyncStatus = when (this) {
    SyncStatus.PENDING -> EntitySyncStatus.PENDING
    SyncStatus.PROCESSING -> EntitySyncStatus.PROCESSING
    SyncStatus.COMPLETED -> EntitySyncStatus.COMPLETED
    SyncStatus.FAILED -> EntitySyncStatus.FAILED
}

File: apps/ios/ScribbleFit/ScribbleFit/Features/AI/Data/Repository/SyncRepositoryImpl.swift
import Foundation
import SwiftData

/**
 * iOS implementation of SyncRepository using ScribbleFitDatabase (SwiftData).
 */
@MainActor
public final class SyncRepositoryImpl: SyncRepository {
    private let database: ScribbleFitDatabase
    private var syncWorkoutUseCase: SyncWorkoutUseCase?
    
    public init(database: ScribbleFitDatabase = .shared) {
        self.database = database
    }
    
    // Lazy injection to avoid circular dependency if any
    public func setSyncWorkoutUseCase(_ useCase: SyncWorkoutUseCase) {
        self.syncWorkoutUseCase = useCase
    }
    
    public func getPendingSyncItems() async throws -> [AISyncItem] {
        return database.getSyncItems(status: .pending).map { $0.toDomain() }
    }
    
    public func getAllSyncItems() async throws -> [AISyncItem] {
        return database.getAllSyncItems().map { $0.toDomain() }
    }
    
    public func updateSyncStatus(id: String, status: AISyncStatus) async throws {
        let dbStatus: ScribbleFit.SyncStatus = switch status {
        case .pending: .pending
        case .processing: .processing
        case .completed: .completed
        case .failed: .failed
        }
        database.updateSyncStatus(id: id, status: dbStatus)
    }
    
    public func saveParsedWorkout(syncItemId: String, workout: ParsedWorkout) async throws {
        database.saveParsedWorkout(syncItemId: syncItemId, workout: workout)
    }
    
    public func enqueueScribble(id: String, rawText: String) async throws {
        let syncItem = SyncQueue(
            id: id,
            itemType: "SCRIBBLE",
            rawText: rawText,
            status: .pending,
            createdAt: Date()
        )
        database.upsertSyncItem(syncItem)
        triggerImmediateSync()
    }
    
    public func saveFeedItem(id: String, itemType: String, jsonData: String, status: AISyncStatus) async throws {
        let dbStatus: ScribbleFit.SyncStatus = switch status {
        case .pending: .pending
        case .processing: .processing
        case .completed: .completed
        case .failed: .failed
        }
        
        let syncItem = SyncQueue(
            id: id,
            itemType: itemType,
            status: dbStatus,
            jsonData: jsonData,
            createdAt: Date()
        )
        database.upsertSyncItem(syncItem)
    }
    
    public func deleteSyncItem(id: String) async throws {
        database.deleteSyncItem(id: id)
    }
    
    private func triggerImmediateSync() {
        Task {
            do {
                try await syncWorkoutUseCase?.execute()
            } catch {
                print("Error executing sync: \(error)")
            }
        }
    }
}

private extension SyncQueue {
    func toDomain() -> AISyncItem {
        let dbStatus = ScribbleFit.SyncStatus(rawValue: self.status) ?? .pending
        let domainStatus: AISyncStatus = switch dbStatus {
        case .pending: .pending
        case .processing: .processing
        case .completed: .completed
        case .failed: .failed
        }
        
        return AISyncItem(
            id: self.id,
            itemType: self.itemType,
            rawText: self.rawText,
            status: domainStatus,
            jsonData: self.jsonData,
            createdAt: self.createdAt
        )
    }
}
